analysis/filter/StructureModifier/src/main/kotlin/de/maibornwolff/codecharta/filter/structuremodifier/FolderMover.kt

```diff
@@ -0,0 +1,97 @@
+package de.maibornwolff.codecharta.filter.structuremodifier
+
+import de.maibornwolff.codecharta.model.*
+import mu.KotlinLogging
+
+class FolderMover(private val project: Project, private val projectName: String?) {
+
+    private val logger = KotlinLogging.logger { }
+    private var toMove: List<MutableNode>? = null
+
+    fun move(moveFrom: String?, moveTo: String?): Project? {
+        if ((moveFrom == null) || (moveTo == null)) {
+            logger.error("In order to move nodes, both source and destination need to be set.")
+            return null
+        }
+
+        return ProjectBuilder(
+                projectName ?: project.projectName,
+                moveNodes(moveFrom, moveTo),
+                extractEdges(moveFrom, moveTo),
+                copyAttributeTypes(),
+                copyBlacklist(moveFrom, moveTo)
+        ).build()
+    }
+
+    private fun getPathSegments(path: String): List<String> {
+        return path.removePrefix("/").split("/")
+    }
+
+    private fun moveNodes(moveFrom: String, moveTo: String): List<MutableNode> {
+        val originPath = getPathSegments(moveFrom)
+        val destinationPath = getPathSegments(moveTo)
+        val rootNode = project.rootNode.toMutableNode()
+
+        val newStructure = listOf(removeMovedNodeFromStructure(originPath, rootNode)!!)
+
+        return if (toMove == null) {
+            logger.warn("Path to move was not found in project. No nodes are therefore moved")
+            newStructure
+        } else {
+            insertInNewStructure(destinationPath.drop(1), rootNode)
+            newStructure
+        }
+    }
+
+    private fun removeMovedNodeFromStructure(originPath: List<String>, node: MutableNode): MutableNode? {
+
+        return if (originPath.isEmpty() || originPath.first() != node.name) {
+            node
+        } else if (originPath.size == 1) {
+            toMove = node.children
+            null
+        } else {
+            node.children = node.children.mapNotNull { child -> removeMovedNodeFromStructure(originPath.drop(1), child) }.toMutableList()
+            node
+        }
+    }
+
+    private fun insertInNewStructure(destinationPath: List<String>, node: MutableNode) {
+        if (destinationPath.isEmpty()) {
+            node.children.addAll(toMove!!)
```

Comment:
What happens with children that that have the same name and the same type? Because we can not import cc.json files with invalid children into the visualisation.

Reply:
that sounds like a case that could occur; i added a filter to sort the duplicate nodes out and give a warning